### PR TITLE
[UniValue] replace global function find_value() with UniValue::findValue()

### PR DESCRIPTION
**Pull request by jonasschnelli**
The global function `find_value()` was implemented in `UniValue` to allow compatibility with JsonSpirit. This is no longer required and we should therefore move it to the UniValue object context.

### CODE REVIEW

**Comment by laanwj** [comment]
Agree on the concept, but this diff is huge. I'd rather postpone it a bit to make sure other pulls don't suffer.

**Comment by jonasschnelli** [comment]
Merging this would produce trivial rebases because the change is very simple `s/find_value(<obj>, <key>)/<obj>.findValue(<key>)/`. Even if new code in PRs use `find_value()`, it's very trivial.

But also no rush in merging this, i can keep this rebased until there is a point in time where not much is going on.

**Comment by jgarzik** [comment]
ut ACK

File: src/univalue/univalue.cpp [hunk 1]
```diff
@@ -164,6 +164,19 @@ int UniValue::findKey(const std::string& key) const
     return -1;
 }
 
+const UniValue UniValue::findValue(const std::string& key) const
+{
+    if (typ != VOBJ)
+        return NullUniValue;
```
Comment (dcousens) [inline src/univalue/univalue.cpp]: Why is this different than the original?

**Comment by dcousens** [comment]
utACK

File: src/univalue/univalue.cpp (hunk 1 above)
Comment (dcousens) [inline src/univalue/univalue.cpp]: This is the only real non-trivial change,  it now defines the behaviour more specifically for non-`VOBJ`.
Wouldn't this have worked already as `keys` would be empty?

Comment (jonasschnelli) [inline src/univalue/univalue.cpp]: Agreed. This check (`if (typ != VOBJ)`) is not really necessary...

Comment (laanwj) [inline src/univalue/univalue.cpp]: Is it a bug if this happens? I'd say `findValue` shouldn't be called for non-VOBJ in the first place. This is similar to calling get_bool on a non-VBOOL. So maybe it should also raise a std::runtime_error.

Comment (laanwj) [inline src/univalue/univalue.cpp]: (but separating this from the trivial change makes sense)

Comment (dcousens) [inline src/univalue/univalue.cpp]: I'd say throwing `std::runtime_error` would be fine,  but agreed,  separation from the trivial change should allow for merge.

**Comment by jonasschnelli** [comment]
Removed the `VOBJ` check, only trivial changes now.

File: src/univalue/univalue.h [hunk 2]
```diff
@@ -154,7 +154,8 @@ class UniValue {
     bool push_back(std::pair<std::string,UniValue> pear) {
         return pushKV(pear.first, pear.second);
     }
-    friend const UniValue& find_value( const UniValue& obj, const std::string& name);
+    //!return a value if key was found (only supports VOBJ types)
```
Comment (dcousens) [inline src/univalue/univalue.h]: is this comment out of date now?

**Comment by dcousens** [comment]
utACK

File: src/univalue/univalue.h (hunk 2 above)
Comment (jonasschnelli) [inline src/univalue/univalue.h]: I would say this comment is appropriate. The keys vector is only >0 for VOBJ objects.

Comment (dcousens) [inline src/univalue/univalue.h]: I feel like the appropriate comment would be `only useful for VOBJ types`,  as technically the function is still "supported".
A nit.

**Comment by jonasschnelli** [comment]
Rebased.

**Comment by dcousens** [comment]
utACK

**Comment by jonasschnelli** [comment]
Rebased.
This should be easy to review.

The function itself is move-only.

The rest of the code just follows search and replace of `s/find_value(<obj>, <key>)/<obj>.findValue(<key>)`.

**Comment by MarcoFalke** [comment]
@jonasschnelli travis complains in case you didn't notice:
```
httprpc.cpp: In function ‘void JSONErrorReply(HTTPRequest*, const UniValue&, const UniValue&)’:

httprpc.cpp:62:43: error: ‘find_value’ was not declared in this scope

**Comment by jonasschnelli** [comment]
@MarcoFalke: Thanks!
It was a rebase issue. Just fixed.

**Comment by jgarzik** [comment]
Strongly in favor of this, but due to size of diff, this should go in at a scheduled point when other "tree stirring" occurs:
* right after univalue subtree merge
* during a lot of libconsensus refactor merges
etc.

**Comment by jonasschnelli** [comment]
This needs to go upstream now. Might come down again when we upgrade UniValue in this repo.
Closing.

**Pull request closed without merging**